planetiler-core/src/main/java/com/onthegomap/planetiler/util/LoopLineMerger.java

```diff
@@ -0,0 +1,339 @@
+import org.locationtech.jts.geom.CoordinateXY;
+import org.locationtech.jts.geom.Geometry;
+import org.locationtech.jts.geom.GeometryComponentFilter;
+import org.locationtech.jts.geom.GeometryFactory;
+import org.locationtech.jts.geom.LineString;
+import org.locationtech.jts.geom.PrecisionModel;
+
+public class LoopLineMerger {
+  private final List<LineString> input = new ArrayList<>();
+  private final List<Node> output = new ArrayList<>();
+  int ids = 0;
+  private PrecisionModel precisionModel = new PrecisionModel(16);
+  private GeometryFactory factory = new GeometryFactory(precisionModel);
+  private double minLength = 0.0;
+  private double loopMinLength = 0.0;
+
+  private static double getLength(List<Edge> lines) {
+    double result = 0.0;
+    for (var line : lines) {
+      result += line.length;
+    }
+    return result;
+  }
+
+  static boolean hasPointAppearingMoreThanTwice(List<Edge> edges) {
+    HashMap<Node, Integer> pointCountMap = new HashMap<>();
+    for (Edge edge : edges) {
+      if (pointCountMap.merge(edge.from, 1, Integer::sum) > 2) {
+        return true;
+      }
+      if (pointCountMap.merge(edge.to, 1, Integer::sum) > 2) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public LoopLineMerger setPrecisionModel(PrecisionModel precisionModel) {
+    this.precisionModel = precisionModel;
+    factory = new GeometryFactory(precisionModel);
+    return this;
+  }
+
+  public LoopLineMerger setMinLength(double minLength) {
+    this.minLength = minLength;
+    return this;
+  }
+
+  public LoopLineMerger setLoopMinLength(double loopMinLength) {
+    this.loopMinLength = loopMinLength;
+    return this;
+  }
+
+  public void add(Geometry geometry) {
+    geometry.apply((GeometryComponentFilter) component -> {
+      if (component instanceof LineString lineString) {
+        input.add(lineString);
+      }
+    });
+  }
+
+  private void merge() {
+    for (var node : output) {
+      if (node.getEdges().size() == 2) {
+        Edge a = node.getEdges().getFirst();
+        Edge b = node.getEdges().get(1);
+        node.getEdges().clear();
+        List<Coordinate> coordinates = new ArrayList<>();
+        coordinates.addAll(a.coordinates.reversed());
+        coordinates.addAll(b.coordinates.subList(1, b.coordinates.size()));
+        Edge c = new Edge(a.to, b.to, coordinates, a.length + b.length);
+        a.to.removeEdge(a.reversed);
+        b.to.removeEdge(b.reversed);
+        a.to.addEdge(c);
+        b.to.addEdge(c.reversed);
+      }
+    }
+  }
+
+  private void removeLoops() {
+    for (var node : output) {
+      for (var edge : List.copyOf(node.getEdges())) {
+        var allPaths = findAllPaths(edge.from, edge.to, loopMinLength);
+        if (allPaths.size() > 1) {
+          for (var path : allPaths.subList(1, allPaths.size())) {
+            if (path.size() > 0) {
+              path.get(0).remove();
+            }
+          }
+        }
+      }
+    }
+  }
```

Comment:
We are removing too many edges here now, since `for (var edge : List.copyOf(node.getEdges())) {` makes a copy of a node's edges at the start, but inside this loop we are removing edges from the graph and it can happen that we remove an edge and later we still check for connects of a removed edge's end nodes... Let me add a small check here...